### ISSUE TITLE
add hashing functions to the `hash` command

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#7817]
hellord :wave:

### Related problem

i simply feel there are soo many hash functions in the wild it's a bit a shame `nushell` only has `md5` and `sha256` :wink: 

it could be quite cool to have a bunch of other hash functions, to enhance and complete the `hash` command a bit :yum: 

### Describe the solution you'd like

> **Note**  
> see the table below for a full list of possible hash functions

> **Warning**  
> i do NOT want nor ask to implement all of them, this list is only there to show what could be added to `hash` :relieved: 

well i do not know what function could be added to `nushell`, and i do not know either where to do that in the source base of `nushell`...

but i'd be happy to have more information and be able to help in that endeavour :muscle: 

cheers :tada: :yum: :wave: 

### Additional context and details

as i'm currenly working on IPFS, i'm quite familiar with the [`multiformats`](https://github.com/multiformats) project :yum: 
over there, they define a bunch of formats, in particular a list of hash functions that could be used inside IPFS without any problem thanks to the standard they define

below is the list of all these hash functions

> **Note**  
> the table has been generated with
> ```bash
> fetch https://raw.githubusercontent.com/multiformats/multicodec/master/table.csv --raw
> | str replace ", *" "," --all
> | from csv
> | where tag == multihash
> | select name code status description
> | first 20
> | to md --pretty
> ```

| name           | code | status    | description                                                                       |
| -------------- | ---- | --------- | --------------------------------------------------------------------------------- |
| identity       | 0x00 | permanent | raw binary                                                                        |
| sha1           | 0x11 | permanent |                                                                                   |
| sha2-256       | 0x12 | permanent |                                                                                   |
| sha2-512       | 0x13 | permanent |                                                                                   |
| sha3-512       | 0x14 | permanent |                                                                                   |
| sha3-384       | 0x15 | permanent |                                                                                   |
| sha3-256       | 0x16 | permanent |                                                                                   |
| sha3-224       | 0x17 | permanent |                                                                                   |
| shake-128      | 0x18 | draft     |                                                                                   |
| shake-256      | 0x19 | draft     |                                                                                   |
| keccak-224     | 0x1a | draft     | keccak has variable output length. The number specifies the core length           |
| keccak-256     | 0x1b | draft     |                                                                                   |
| keccak-384     | 0x1c | draft     |                                                                                   |
| keccak-512     | 0x1d | draft     |                                                                                   |
| blake3         | 0x1e | draft     | BLAKE3 has a default 32 byte output length. The maximum length is (2^64)-1 bytes. |
| sha2-384       | 0x20 | permanent | aka SHA-384; as specified by FIPS 180-4.                                          |
| murmur3-x64-64 | 0x22 | permanent | The first 64-bits of a murmur3-x64-128 - used for UnixFS directory sharding.      |
| murmur3-32     | 0x23 | draft     |                                                                                   |
| dbl-sha2-256   | 0x56 | draft     |                                                                                   |
| md4            | 0xd4 | draft     |                                                                                   |

**Comment by amtoine** [comment]
maybe a plugin could be great for this? :open_mouth: 
you tell me what you'd prefer :wink:

**Comment by fdncred** [comment]
> maybe a plugin could be great for this? 

Yup, that's exactly what I was thinking. You can have 1 plugin with every single one of these in it if you wanted. Plugins support a vec of signatures which allows you to combine many things into one plugin. I'd say go for it.

**Comment by amtoine** [comment]
> Yup, that's exactly what I was thinking. You can have 1 plugin with every single one of these in it if you wanted. Plugins support a vec of signatures which allows you to combine many things into one plugin. I'd say go for it.
>
noice :ok_hand: 
a good way to start on plugins, after the `str len` example of the book :smirk:

**Comment by jwillikers** [comment]
I would appreciate there being more hash options built-in, too. Namely SHA-1 and SHA3-512.

**Comment by fdncred** [comment]
Yup, more hash options would be great. Someone just needs to create a plugin and add them.

**Comment by amtoine** [comment]
:+1:

**Comment by vlovgr** [comment]
Since a month or so back, there is now [nu_plugin_hashes](https://github.com/ArmoredPony/nu_plugin_hashes) by @ArmoredPony. 🎉 

Is there any interest in having this plugin as part of the Nushell repository?

**Comment by fdncred** [comment]
Probably not. I doubt the core-team is interested in maintaining 61 flavors of hashing.

**Comment by vlovgr** [comment]
> Probably not. I doubt the core-team is interested in maintaining 61 flavors of hashing.

That makes a lot of sense. 👍 Just to add my two cents, but I think there might be a reasonable middle-ground between providing 2 hashing functions (as of today), and having to resort to a third-party plugin. But if the core team is happy with the current state, then I believe this issue can be closed?

**Comment by fdncred** [comment]
Agree about middle-ground. The problem is, if I ask 100 developers, "What are the most popular hashing algorithms", I'll probably get a bunch of different answers. Where do you feel the middle-ground is?

**Comment by ArmoredPony** [comment]
nushell already has support for algorithms that implement `Digest` trait. I just copied the code, pasted it into my plugin and generated actual nushell commands using a build script. I think they can be handled no differently from two built-in hashing algorithms

**Comment by sholderbach** [comment]
As the plugin exists and we generally endorse the use of plugins for those cases where it is less clear which functionality is paramoutn for a broad userbase, we will close this issue and refer to `nu_plugin_hashes` or similar implementations. If the need for a particular function would benefit the general userbase we can open a specific issue and discuss there.

Thanks @ArmoredPony for your work!
